### PR TITLE
fix(minibf): fix ordering in account addresses

### LINKED ISSUE #1140: minibf: fix `/accounts/{stake_address}/addresses`
Fix implementation of the Blockfrost `/accounts/<account>/addresses` endpoint in the `minibf` crate.

Bug: edge cases ignored in `blockfrost-tests`: address ordering (mainnet), zero-address account (preprod).

### Effort

Unestimated

### Approach

Unknown

### Why now

0.3172% of observed traffic.

### PR DESCRIPTION
**Pull request by slowbackspace**
Closes #1140.

## Summary

Fixes the `desc` ordering of `GET /accounts/{stake_address}/addresses` and adds a stake address log to the archive store, so the endpoint answers every page in O(page) time.

## Bug: `order=desc` returned the wrong order

- **Cause**: the handler scanned newest-first and placed each reused address at its latest appearance. Blockfrost orders by first appearance (ryo's `accounts_stake_address_addresses.sql`).
- **Fix**: both orders read the stake address log, which stores first appearances. `desc` is a reverse range read, so it is the exact reverse of `asc` by construction.
- **Proof**: all four mainnet fixture cases pass on a synced mainnet snapshot; the `desc` case failed before.

## Stake address log

No scan can serve this endpoint. A scan must read the account's whole history to prove it found every address. Measured on a preview store copy (five heaviest accounts, `count=100`, seconds):

| Account | Tagged blocks | Distinct addresses | `asc` scan | `desc` scan (asc + reverse) | log, either order (this PR) |
|---|---|---|---|---|---|
| `stake_test1uz2a3lk22…` | 821k | 4 | 226 | 226 | 0.0004–0.002 |
| `stake_test1ur96gg3s3…` | 185k | 200+ | 0.07 | 20–21 | 0.0004–0.002 |
| `stake_test1upv7n2x0l…` | 160k | 200+ | 0.02 | 15.2–15.4 | 0.0004–0.002 |
| `stake_test1urk948umh…` | 147k | 3 | 15.2–15.6 | 14.9–15.7 | 0.0004–0.002 |
| `stake_test1uzd7tg4e6…` | 56k | 1 | 8.2 | 7.3–7.9 | 0.0004–0.002 |

The old scan took 305 seconds for a mainnet exchange account with 400k+ addresses — a DoS shape on public nodes. So the endpoint reads the log, and only the log.

The log stores each `(stake credential, address)` pair once, at its first on-chain appearance, ordered by `(slot, tx order, output order)`.

| Piece | Where | Shape |
|---|---|---|
| Delta | `ArchiveIndexDelta::stake_addresses` | one `StakeAddressAppearance { order, stake, address }` per produced output with a stake credential |
| Write | `ArchiveWriter::apply_index` / `undo_index` | same call and same batch as the archive tags and exact lookups |
| Read | `ArchiveStore::addresses_by_stake_log(stake, offset, limit, reverse)` | one page, in either order |

- **fjall**: new `archive-stake-log` keyspace with two entry shapes: a pair entry per `(stake, address)` for the write-path probe and undo, and an ordered entry for page reads. The writer keeps a pending map of the pairs it inserted or removed, so apply and undo in one writer see each other's writes. Not swept by `prune_history` — entries are first appearances, so pruning one would drop an address the account still uses.
- **memory**: implements the same semantics, so `ToyDomain` endpoint tests exercise the log; **noop** answers an empty page.
- **Undo**: rollback rebuilds the deltas with the same `index_block` the apply path uses, so `undo_index` removes exactly what `apply_index` inserted. A pair is removed only when the undone block is its stored first appearance.
- **Population**: relay sync, Mithril bootstrap, WAL catch-up and snapshot backfill write the log in the same batch as the blocks.
- **Migration**: `doctor rebuild-state --rewrite-logs` backfills existing stores. Its gate used to drop `apply_index` whole, so a rebuild computed the entries and discarded them; the gate now forwards only the `stake_addresses` field, and the backend skips the emptied tag and exact inputs, so the rest of the archive stays read-only. Until the rebuild or a resync runs, an upgraded store answers with the addresses seen since the upgrade only. A replay that reaches an appearance earlier than the stored one moves the pair to it, so the rebuild corrects a log that started mid-chain; re-running over a complete log changes nothing. A resync works too.
- **Known gap, stelae**: a stele restore does not write the log — the `indexes` stele layer carries tag and exact records only. Closing that needs a third record kind and a `PROFILE_VERSION` bump; follow-up on the stele profile (#1217). Until then a stele-bootstrapped node backfills with `doctor rebuild-state --rewrite-logs`, which reads the restored archive.
- **Size**: full address bytes are stored — the payment part is not derivable from the stake credential, and a page must return each address exactly as it appeared; estimated 5–15 GB on mainnet against a 340 GB store. Pointer addresses never enter the log: `shelley_address_to_stake_address` resolves no pointer, so dolos attributes them to no account anywhere.

## Testing

- `accounts_by_stake_addresses_order_desc`: `desc` equals the reverse of `asc`, and a `desc` page is a window into the reversed list.
- `stake_log_round_trips_and_pages` (fjall and memory): first-appearance dedup, ordered paging from both ends, offset windows, in-block ordering, stake isolation, undo of the first appearance only.
- `stake_log_replay_moves_a_pair_to_an_earlier_appearance` (fjall and memory): a pair recorded at a later slot moves when a replay finds it earlier, no stale entry remains, undo follows the new position, and a later appearance never moves a pair. Fails on both backends without the fix.
- `stake_log_apply_and_undo_share_one_writer` (fjall and memory): apply then undo, and undo then apply, of one block inside one writer leave the log as separate commits would. Failed on fjall before the fix: its `undo` read only the committed snapshot and missed the writer's own inserts.
- `logs_only_gate_writes_the_stake_log_and_nothing_else`: after an `apply_index` through the `--rewrite-logs` gate, the log answers and the same delta's tag and exact entries stay unwritten.
- Live verification on a preview store bootstrapped with this branch: every accounts content fixture passes, answering in 1–2 ms from the log (remaining suite failures are its new `Bf-Tag`/`kupo_only` routing assertions, which no dolos backend can satisfy). A 63-check live diff against the Blockfrost preview API matched on everything this branch controls; the only divergence is one credential of the pre-existing #1325 class.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added a stake address history log recording each address’s first on-chain appearance for a stake account.
  - Account address lookups now support pagination and forward or reverse first-appearance ordering.
  - Stake address history remains available beyond archive history pruning.
  - Added support across memory, persistent, and gated archive storage.

- **Documentation**
  - Updated architecture and archive storage documentation to describe the stake address log and its retention behavior.

### CODE REVIEW

**Comment by michalrus** [comment]
`cargo-deny` fixed in:
- #1229

**Comment by slowbackspace** [comment]
Verification on a preview store bootstrapped with this branch (merged locally with main + #1320 for the test):

- Accounts fixture family: every content fixture passes, including the previously heavy ones — `addresses` for the many-address account, dormant single-address, zero-address, and desc ordering, each answering in 1-2 ms from the stake log. The only failures are suite routing artifacts (`Bf-Tag` header, `kupo_only` twins) and fixtures for endpoints not in this branch.
- Live diff vs the Blockfrost preview API: 63 checks over 33 real accounts (proposal return addresses across the chain plus the fixture shapes) with desc/pagination/total/assets variants and negative probes — 59/59 parity on everything this branch controls. The four mismatches are all one credential and are the pre-existing #1325 divergence (proposal-return-only account: dolos 404, BF `200 []`).
- Votes fixtures still pass on the same store (the branch coexists with #1320's state changes).

One finding that needs a decision before undrafting: the 32ce4c3b commit message claims `doctor rebuild-state` populates the stake address log — it does not. The log rides in `ArchiveIndexDelta.stake_addresses` through the archive writer's `apply_index`, which both the default no-op archive and the `--rewrite-logs` `LogsOnly` gate discard (storage.rs `Self::LogsOnly(_) => Ok(())`). Verified empirically: after a full in-place rebuild the log keyspace exists and is empty, and the addresses endpoints return `[]` for every account. Either widen the `LogsOnly` gate to pass the stake-address part of `apply_index` (the replay already computes it every run; blocks/tags/exact stay untouched), or change the message and docs to resync-only — today the endpoint fails silently on every upgraded store.

Rebase note: current main's `crates/core/tests/block_meta.rs` constructs `ArchiveIndexDelta` without the new field — four sites need `stake_addresses: Vec::new()` on rebase.

File: crates/minibf/src/routes/accounts.rs [hunk 1]
```diff
@@ -224,79 +224,111 @@ where
...
+        // - `live` is the only slot a brand-new pool writes.
+        // - `next` holds a mid-epoch re-registration until the boundary.
+        // - `mark`/`set`/`go` still hold a credential that a recent re-registration
+        //   replaced.
+        let snapshots = [
+            pool.snapshot.live(),
+            pool.snapshot.next(),
+            pool.snapshot.mark(),
+            pool.snapshot.set(),
+            pool.snapshot.go(),
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply crates/minibf/src/routes/accounts.rs]: This fallback was split out to #1339; the rotated-credential gap is tracked in #1325. Not this PR's scope.

File: crates/minibf/src/routes/accounts.rs [hunk 2]
```diff
@@ -224,79 +224,111 @@ where
...
-            continue;
-        };
+    // Every page is one read of the stake address log, in either order. The
+    // log is written in the same batch as the blocks it projects, by every
+    // path that applies blocks from genesis.
+    let addresses = domain
+        .archive()
+        .addresses_by_stake_log(
+            &account_key.address.to_vec(),
+            pagination.skip(),
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply crates/minibf/src/routes/accounts.rs:254]: Documented in the description as the known stelae gap. Two facts narrow it: the `indexes` stele layer carries tag and exact records only, so the fix is a third record kind plus a profile version bump — a stele-profile follow-up, not this PR. And since 881047cd on this branch, `doctor rebuild-state --rewrite-logs` backfills the log from the restored archive, so a stele-bootstrapped node has an offline path to a complete log without a resync. The completeness marker was removed in 32ce4c3b on purpose: the fallback it gated was the 226-second scan the log replaces, so the endpoint either has the index or is not worth serving from a scan.

File: crates/minibf/src/routes/accounts.rs [hunk 3]
```diff
@@ -224,79 +224,111 @@ where
...
+
+    Ok(false)
+}
+
+/// `GET /accounts/{stake_address}/addresses`.
+///
+/// Blockfrost declares only `count`, `page` and `order` for this endpoint.
+/// The shared query struct also accepts `from` and `to`; they are ignored
+/// here, as Blockfrost ignores them, so the list always covers the account's
+/// whole history.
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply crates/minibf/src/routes/accounts.rs]: `from`/`to` handling was split out to #1339 and the malformed-value case is fixed there. Not this PR's scope.

File: crates/fjall/src/archive/README.md [hunk 4]
```diff
@@ -54,6 +55,21 @@ Value: (empty)
...
+| Entry | Key | Value |
+|-------|-----|-------|
+| Pair | `[0x00][stake_len:1][stake][address]` | `[slot:8][order:4]` |
+| Ordered | `[0x01][stake_len:1][stake][slot:8][order:4][address]` | (empty) |
+
+Each `(stake, address)` pair is stored once, at its first on-chain appearance;
+`order` is the transaction index in the high 16 bits and the output index in
+the low 16. The pair entry is the write-path probe and the undo key; the
+ordered entry is what a page read walks, from either end. Full address bytes
+are stored so pointer addresses round-trip.
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply crates/fjall/src/archive/README.md]: Correct — `shelley_address_to_stake_address` resolves no pointer, so pointer addresses never reach the log. Reworded in 632971e3: the entry stores full address bytes because the payment part is not derivable from the stake credential and a page must return each address exactly as it appeared. The pointer limitation is not new to the log: the same helper feeds the archive `STAKE` tag, so dolos attributes pointer addresses to no account anywhere.

File: crates/fjall/src/archive/mod.rs [hunk 5]
```diff
@@ -573,6 +609,18 @@ impl CoreArchiveWriter for ArchiveWriter {
...
 
+        let snapshot = self.store.db.snapshot();
+
+        for block in deltas.iter().rev() {
+            stake_log::undo(
+                &mut batch,
+                &self.store.stake_log,
+                &snapshot,
+                block.slot,
+                &block.stake_addresses,
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply crates/fjall/src/archive/mod.rs:622]: Confirmed and fixed in edf08eae. A new conformance test, `stake_log_apply_and_undo_share_one_writer`, runs apply then undo and undo then apply of one block inside one writer; it failed on fjall (the pair stayed behind) and passed on memory before the fix.

The apply-only `seen` set is now a per-writer pending map of positions (`Some` after an insert, `None` after a removal). Both `apply` and `undo` consult it before the committed snapshot, so a writer sees its own writes the way the memory backend does. No production writer mixes the two calls today — the forward batch and the bootstrap only apply, `rollback` only undoes — so this closes a latent gap rather than a live one.

File: src/adapters/storage.rs [hunk 6]
```diff
@@ -1579,4 +1625,52 @@ mod lifecycle_tests {
...
+        let page = store.addresses_by_stake_log(&stake, 0, 10, false).unwrap();
+        assert_eq!(page, vec![address]);
+
+        // the tag and exact entries the delta carried were not written
+        let slots: Vec<u64> = store
+            .slots_by_tag("account_certs", &[1, 2, 3], 0, u64::MAX)
+            .unwrap()
+            .collect::<Result<_, _>>()
+            .unwrap();
+        assert!(slots.is_empty());
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply src/adapters/storage.rs]: Confirmed and fixed in 8308cc11. The test now also asserts `slot_by_block_hash`, `slot_by_block_number` and `slot_by_tx_hash` all miss for this delta.

Verified the assertions bite: forwarding the delta's `block_hash`, `block_number` and `tx_hashes` through the logs-only projection instead of emptying them fails the test at the first of the three.
